common/context/src/main/java/io/helidon/common/context/ContextValue.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.common.context;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+/**
+ * This is similar to {@link java.lang.ThreadLocal}, except it uses context to store a value.
+ * <p>
+ * When test context is available as a parent of current context
+ * (identified as a context that has ID starting with {@code test-}), the instance is registered with that context.
+ * Otherwise, it is registered with {@link io.helidon.common.context.Contexts#globalContext()}.
+ * <p>
+ * The intention of this type is to have singletons for production runtime, while maintaining a separation during
+ * testing, where Helidon test extension run each part of the test class in a test context.
+ * <p>
+ * Note that the instance is not simply registered in the context, but it uses this class as a classifier.
+ *
+ * @param <T> type of the value stored in context
+ */
+public final class ContextValue<T> {
```

Comment:
The name `ContextValue` is ambiguous.

I would expect `ContextValue` to follow the same pattern as `ConfigValue` / `OptionalValue` and replace `Optional` as the return type for the methods of `Context`.

Maybe that's not a bad idea, but that's out of scope for this PR.

---

This class is about singletons or "static values", the name and the context qualifier should reflect that.
The current qualifier is `global-instances`, but `Global` is already overloaded with `Context.globalContext()`.

I like the term `static` because it's short, however I'm not convinced with either `StaticValue` or `ContextStaticValue`. `ContextSingleton` does not seem too bad.

---

Another solution could be to separate the "mutable value" and the "static context". 

I.e. A generic `MutableValue`, similar to `LazyValue`, combined with `Contexts.staticContext()`.

E.g.

```java
static final MutableValue<Config> CONFIG_VALUE = MutableConfig.create(
    Contexts::staticContext,
    (c, v) -> c.set(Config.class, v),
    c -> c.get(Config.class).orElseThrow());
```

---

You stated earlier that one of the reasons we can't use service registry currently is because the global values are mutable.

Could we implement registry mutability with a special wrapper type like `MutableValue` ?

```java
static GenericType<MutableValue<Config>> MUTABLE_TYPE = new GenericType<MutableValue<Config>>() {};
static LazyValue<Config> DEFAULT_CONFIG = LazyValue.create(Config::empty);

static Config config() {
    return GlobalServiceRegistry.registry()
            .first(MUTABLE_TYPE)
            .map(MutableValue::get)
            .orElseGet(DEFAULT_CONFIG);
}
```

`GlobalServiceRegistry.registry()` would be based on a "static context".

Reply:
Problem with Registry is also cyclic dependency - Registry must depend on config, and config already uses a "global" instance concep.

Reply:
Same with metrics - registry depends on metrics API

Reply:
Agreed on the naming. Will try a different one.

Reply:
Tried a different name, please re-check (also using a constant for the context registered)